File: algorithms/dijkstra.java
// Dijkstra's Algorithm is used to find the shortest path in a graph with non-negative edge weights.

// It works by maintaining a priority queue of nodes whose shortest path from the source node has been determined.
// It begins by setting the distance to the source node to zero and all other nodes to infinity (or a very large value).
// It repeatedly selects the node with the smallest distance from the priority queue, relaxes its neighboring nodes by updating their distances if a shorter path is found, and continues until all nodes have been processed.

// Imagine a road network where cities are nodes, and roads between cities are edges with weights representing distances.
// Dijkstra's algorithm can be used to find the shortest path (minimum distance) between two cities.

import java.util.Comparator;
import java.util.HashMap;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;

class Graph {
    private Map<Integer, List<int[]>> adjList;

    public Graph() {
        this.adjList = new HashMap<>();
    }

    public void addEdge(int u, int v, int weight) {
        adjList.putIfAbsent(u, new ArrayList<>());
        adjList.get(u).add(new int[](v, weight));
    }

    public Map<Integer, Integer> dijkstra(int source) {
        Map<Integer, Integer> distances = new HashMap<>();
        PriorityQueue<int[]> minHeap = new PriorityQueue<>(Comparator.comparingInt(a -> a[1]));

        // Init distances with infinity
        for (int node : adjList.keySet()) {
            distances.put(node, Integer.MAX_VALUE);
        }

        distances.put(source, 0);
        minHeap.offer(new int[]{source, 0});

        while (!minHeap.isEmpty()) {
            int[] curr = minHeap.poll();
            int currNode = curr[0];
            int currDist = curr[1];

            if (currDist > distance.get(currNode)) {
                continue;
            }

            // Explore neighbors
            if (adjList.containsKey(currNode)) {
                for (int[] neighbor : adjList.get(currNode)) {
                    int nextNode = neighbor[0];
                    int edgeWeight = neighbor[1];
                    int newDist = currDist + edgeWeight;

                    // Update distance if a shorter path is found
                    if (newDist < distances.get(nextNode)) {
                        distance.put(nextNode, newDist);
                        minHeap.offer(new int[]{nextNode, newDist});
                    }
                }
            }
        }

        return distances;
    }
}

File: algorithms/dijkstra.kt
// Dijkstra's Algorithm is a graph search algorithm that solves the single-source shortest path problem for a graph with non-negative edge weights, producing the shortest path tree.
// This algorithm is often used in routing and as a subroutine in other graph algorithms.

// Imagine you're using a GPS navigation system to find the shortest path from your home to a destination in a city.
// The streets form a graph, and the driving time between intersections represent weights.
// Dijkstra's Algorithm can be used to determine the shortest time route to your destination, considering all possible paths.

data class Edge(val to: Int, val weight: Int)

fun dijkstra(graph: List<List<Edge>>, start: Int): IntArray {
    val n = graph.size
    val dist = IntArray(n) { Int.MAX_VALUE } // Distance values from the start
    val visited = BooleanArray(n) { false }

    dist[start] = 0

    for (i in 0 until n) {
        // Find the vertex with the minimum distance value from the unvisited vertices
        val u = (0 until n).filter { !visited[it] }.minByOrNull { dist[it] } ?: break

        // Mark the vertex as visited
        visited[u] = true

        // Update the distance value of the adjacent vertices
        for (edge in graph[u]) {
            val v = edge.to
            if (!visited[v] && dist[u] != Int.MAX_VALUE && dist[u] + edge.weight < dist[v]) {
                dist[v] = dist[u] + edge.weight
            }
        }
    }

    return dist
}

fun main() {
    val graph = listOf(
        listOf(Edge(1, 4), Edge(2, 2)), // Neighbors of node 0
        listOf(Edge(2, 5), Edge(3, 2)), // Neighbors of node 1
        listOf(Edge(3, 3)), // Neighbors of node 2
        listOf()
    )

    val distancesFromNode0 = dijkstra(graph, 0)
    println(distancesFromNode0.joinToString(", ")) // Output: 0, 4, 2, 5
}